Panic executing: SELECT author_name from commits order by COUNT(*);

```
--> Executing query: SELECT author_name from commits order by COUNT(*);

panic: interface conversion: interface is string, not int32

goroutine 1 [running]:
panic(0x9839e0, 0xc420fb87c0)
	/usr/local/go/src/runtime/panic.go:500 +0x1a1
github.com/gitql/gitql/sql.(*integerType).Compare(0xd5d788, 0x949d20, 0xc4201ef910, 0x949d20, 0xc429cec240, 0xc429cec240)
	<autogenerated>:45 +0x82
github.com/gitql/gitql/sql/plan.(*sorter).Less(0xc42811bf20, 0x0, 0x1792, 0x0)
	/home/antonio/work/src/github.com/gitql/gitql/sql/plan/sort.go:156 +0x1e5
sort.medianOfThree(0xcf59c0, 0xc42811bf20, 0x0, 0x1792, 0x2f24)
	/usr/local/go/src/sort/sort.go:74 +0x49
sort.doPivot(0xcf59c0, 0xc42811bf20, 0x0, 0xbc97, 0x53a887, 0xc42006c400)
	/usr/local/go/src/sort/sort.go:99 +0x601
sort.quickSort(0xcf59c0, 0xc42811bf20, 0x0, 0xbc97, 0x1f)
	/usr/local/go/src/sort/sort.go:188 +0x83
sort.Sort(0xcf59c0, 0xc42811bf20)
	/usr/local/go/src/sort/sort.go:222 +0x80
github.com/gitql/gitql/sql/plan.(*sortIter).computeSortedRows(0xc4219fe540, 0x1, 0xb)
	/home/antonio/work/src/github.com/gitql/gitql/sql/plan/sort.go:126 +0x292
github.com/gitql/gitql/sql/plan.(*sortIter).Next(0xc4219fe540, 0xc425fc24d0, 0xc4201ef610, 0xc42030d9e0, 0x4a19fc, 0xc425fc2480)
	/home/antonio/work/src/github.com/gitql/gitql/sql/plan/sort.go:92 +0xd9
github.com/gitql/gitql/sql/plan.(*iter).Next(0xc42571f940, 0xc4201ef600, 0x1, 0x1, 0x0, 0x0)
	/home/antonio/work/src/github.com/gitql/gitql/sql/plan/project.go:65 +0x38
main.(*cmdQueryBase).printQuery(0xc4201d6b40, 0xc42571f960, 0x1, 0x1, 0xcf2700, 0xc42571f940, 0xa34a17, 0x6, 0x0, 0x0)
	/home/antonio/work/src/github.com/gitql/gitql/cmd/gitql/query_base.go:65 +0x222
main.(*CmdShell).Execute(0xc4201d6b40, 0xc4201a96c0, 0x0, 0x1, 0x1, 0x1)
	/home/antonio/work/src/github.com/gitql/gitql/cmd/gitql/shell.go:75 +0xa5a
github.com/jessevdk/go-flags.(*Parser).ParseArgs(0xc420022a20, 0xc42000c5f0, 0x1, 0x1, 0x4, 0x2, 0xc42001c900, 0xc4200ab080, 0xc4200aaf48)
	/home/antonio/work/src/github.com/jessevdk/go-flags/parser.go:316 +0x8e6
github.com/jessevdk/go-flags.(*Parser).Parse(0xc420022a20, 0xa36250, 0x7, 0xa44624, 0x1d, 0x0)
	/home/antonio/work/src/github.com/jessevdk/go-flags/parser.go:186 +0x74
main.main()
	/home/antonio/work/src/github.com/gitql/gitql/cmd/gitql/main.go:16 +0x38b
```

--- a/sql/analyzer/analyzer.go
+++ b/sql/analyzer/analyzer.go
@@ -1,7 +1,6 @@
 package analyzer
 
 import (
-	"errors"
 	"fmt"
 	"reflect"
 
@@ -12,6 +11,7 @@ const maxAnalysisIterations = 1000
 
 type Analyzer struct {
 	Rules           []Rule
+	ValidationRules []ValidationRule
 	Catalog         *sql.Catalog
 	CurrentDatabase string
 }
@@ -21,10 +21,16 @@ type Rule struct {
 	Apply func(*Analyzer, sql.Node) sql.Node
 }
 
+type ValidationRule struct {
+	Name  string
+	Apply func(*Analyzer, sql.Node) error
+}
+
 func New(catalog *sql.Catalog) *Analyzer {
 	return &Analyzer{
-		Rules:   DefaultRules,
-		Catalog: catalog,
+		Rules:           DefaultRules,
+		ValidationRules: DefaultValidationRules,
+		Catalog:         catalog,
 	}
 }
 
@@ -41,7 +47,12 @@ func (a *Analyzer) Analyze(n sql.Node) (sql.Node, error) {
 		}
 	}
 
-	return cur, a.validate(cur)
+	// TODO improve error handling
+	if errs := a.validate(cur); len(errs) != 0 {
+		return cur, errs[0]
+	}
+
+	return cur, nil
 }
 
 func (a *Analyzer) analyzeOnce(n sql.Node) sql.Node {
@@ -52,10 +63,23 @@ func (a *Analyzer) analyzeOnce(n sql.Node) sql.Node {
 	return result
 }
 
-func (a *Analyzer) validate(n sql.Node) error {
-	if !n.Resolved() {
-		return errors.New("plan is not resolved")
+func (a *Analyzer) validate(n sql.Node) (validationErrors []error) {
+	validationErrors = append(validationErrors, a.validateOnce(n)...)
+
+	for _, node := range n.Children() {
+		validationErrors = append(validationErrors, a.validate(node)...)
 	}
 
-	return nil
+	return validationErrors
+}
+
+func (a *Analyzer) validateOnce(n sql.Node) (validationErrors []error) {
+	for _, rule := range a.ValidationRules {
+		err := rule.Apply(a, n)
+		if err != nil {
+			validationErrors = append(validationErrors, err)
+		}
+	}
+
+	return validationErrors
 }

--- a/sql/analyzer/validation_rules.go
+++ b/sql/analyzer/validation_rules.go
@@ -1,0 +1,35 @@
+package analyzer
+
+import (
+	"errors"
+
+	"gopkg.in/sqle/sqle.v0/sql"
+	"gopkg.in/sqle/sqle.v0/sql/plan"
+)
+
+var DefaultValidationRules = []ValidationRule{
+	{"validate_resolved", validateIsResolved},
+	{"validate_order_by", validateOrderBy},
+}
+
+func validateIsResolved(a *Analyzer, n sql.Node) error {
+	if !n.Resolved() {
+		return errors.New("plan is not resolved")
+	}
+
+	return nil
+}
+
+func validateOrderBy(a *Analyzer, n sql.Node) error {
+	switch n := n.(type) {
+	case *plan.Sort:
+		for _, field := range n.SortFields {
+			switch field.Column.(type) {
+			case sql.AggregationExpression:
+				return errors.New("OrderBy does not support aggregation expressions")
+			}
+		}
+	}
+
+	return nil
+}

--- a/sql/analyzer/validation_rules_test.go
+++ b/sql/analyzer/validation_rules_test.go
@@ -1,0 +1,66 @@
+package analyzer_test
+
+import (
+	"testing"
+
+	"gopkg.in/sqle/sqle.v0/sql"
+	"gopkg.in/sqle/sqle.v0/sql/analyzer"
+	"gopkg.in/sqle/sqle.v0/sql/expression"
+	"gopkg.in/sqle/sqle.v0/sql/plan"
+
+	"github.com/stretchr/testify/require"
+)
+
+func Test_resolved(t *testing.T) {
+	assert := require.New(t)
+
+	vr := getValidationRule("validate_resolved")
+
+	assert.Equal(vr.Name, "validate_resolved")
+
+	err := vr.Apply(nil, dummyNode{true})
+	assert.NoError(err)
+
+	err = vr.Apply(nil, dummyNode{false})
+	assert.Error(err)
+
+}
+
+func Test_orderBy(t *testing.T) {
+	assert := require.New(t)
+
+	vr := getValidationRule("validate_order_by")
+
+	assert.Equal(vr.Name, "validate_order_by")
+
+	err := vr.Apply(nil, dummyNode{true})
+	assert.NoError(err)
+	err = vr.Apply(nil, dummyNode{false})
+	assert.NoError(err)
+
+	err = vr.Apply(nil, plan.NewSort(
+		[]plan.SortField{{expression.NewCount(nil),
+			plan.Descending}}, nil))
+	assert.Error(err)
+}
+
+type dummyNode struct{ resolved bool }
+
+func (n dummyNode) Resolved() bool                             { return n.resolved }
+func (dummyNode) Schema() sql.Schema                           { return sql.Schema{} }
+func (dummyNode) Children() []sql.Node                         { return nil }
+func (dummyNode) RowIter() (sql.RowIter, error)                { return nil, nil }
+func (dummyNode) TransformUp(func(sql.Node) sql.Node) sql.Node { return nil }
+func (dummyNode) TransformExpressionsUp(
+	func(sql.Expression) sql.Expression) sql.Node {
+	return nil
+}
+
+func getValidationRule(name string) analyzer.ValidationRule {
+	for _, rule := range analyzer.DefaultValidationRules {
+		if rule.Name == name {
+			return rule
+		}
+	}
+	panic("missing rule")
+}

--- a/sql/plan/sort.go
+++ b/sql/plan/sort.go
@@ -9,7 +9,7 @@ import (
 
 type Sort struct {
 	UnaryNode
-	sortFields []SortField
+	SortFields []SortField
 }
 
 type SortOrder byte
@@ -27,7 +27,7 @@ type SortField struct {
 func NewSort(sortFields []SortField, child sql.Node) *Sort {
 	return &Sort{
 		UnaryNode:  UnaryNode{child},
-		sortFields: sortFields,
+		SortFields: sortFields,
 	}
 }
 
@@ -36,7 +36,7 @@ func (s *Sort) Resolved() bool {
 }
 
 func (p *Sort) expressionsResolved() bool {
-	for _, f := range p.sortFields {
+	for _, f := range p.SortFields {
 		if !f.Column.Resolved() {
 			return false
 		}
@@ -55,7 +55,7 @@ func (s *Sort) RowIter() (sql.RowIter, error) {
 
 func (s *Sort) TransformUp(f func(sql.Node) sql.Node) sql.Node {
 	c := s.UnaryNode.Child.TransformUp(f)
-	n := NewSort(s.sortFields, c)
+	n := NewSort(s.SortFields, c)
 
 	return f(n)
 }
@@ -63,7 +63,7 @@ func (s *Sort) TransformUp(f func(sql.Node) sql.Node) sql.Node {
 func (s *Sort) TransformExpressionsUp(f func(sql.Expression) sql.Expression) sql.Node {
 	c := s.UnaryNode.Child.TransformExpressionsUp(f)
 	sfs := []SortField{}
-	for _, sf := range s.sortFields {
+	for _, sf := range s.SortFields {
 		sfs = append(sfs, SortField{sf.Column.TransformUp(f), sf.Order})
 	}
 	n := NewSort(sfs, c)
@@ -121,7 +121,7 @@ func (i *sortIter) computeSortedRows() error {
 		rows = append(rows, childRow)
 	}
 	sort.Sort(&sorter{
-		sortFields: i.s.sortFields,
+		sortFields: i.s.SortFields,
 		rows:       rows,
 	})
 	i.sortedRows = rows

